### PR TITLE
Migration of purchase_request_to_rfq to 10.0

### PR DESCRIPTION
**Pull request by adrienpeiffer**
- [x] Forward modification on purchase_request_to_rfq after merge in 9.0

### CODE REVIEW

File: purchase_request_to_rfq/models/purchase_order.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        'Purchase Request Lines', readonly=True, copy=False)
+
+    @api.multi
+    def action_openRequestLineTreeView(self):
+        """
+        :return dict: dictionary value for created view
+        """
+        request_line_ids = []
+        for line in self:
+            request_line_ids += line.purchase_request_lines.ids
```
Comment (Garamotte) [inline purchase_request_to_rfq/models/purchase_order.py:96]: `request_line_ids = line.mapped('purchase_request_lines').ids`

File: purchase_request_to_rfq/models/purchase_request.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    def _compute_is_editable(self):
+        super(PurchaseRequestLine, self)._compute_is_editable()
+        self.filtered(lambda p: p in self and p.purchase_lines).update(
+            {'is_editable': False})
+
+    @api.multi
+    def _compute_purchased_qty(self):
+        for rec in self:
+            rec.purchased_qty = sum(rec.purchase_lines.filtered(
+                lambda x: x.state != 'cancel').mapped('product_qty'))
```
Comment (Garamotte) [inline purchase_request_to_rfq/models/purchase_request.py]: Use the compute_quantity method to convert all quantities to the same UoM before summing values.

File: purchase_request_to_rfq/models/stock.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    @api.multi
+    def do_transfer(self):
+        super(StockPicking, self).do_transfer()
+        request_obj = self.env['purchase.request']
+        for picking in self:
+            requests_dict = {}
+            if picking.picking_type_id.code != 'incoming':
+                continue
+            for move in picking.move_lines:
+                if move.purchase_line_id:
```
Comment (Garamotte) [inline purchase_request_to_rfq/models/stock.py:43]: The `if` statement is useless here.

File: purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+        po_line_obj = self.env['purchase.order.line']
+        product = item.product_id
+        supplier = self.supplier_id
+        pricelist_id = supplier.property_product_pricelist
+
+        if pricelist_id:
+            date_order_str = datetime.strptime(
+                fields.datetime.now().strftime(DEFAULT_SERVER_DATETIME_FORMAT),
+                DEFAULT_SERVER_DATETIME_FORMAT).\
+                strftime(DEFAULT_SERVER_DATE_FORMAT)
```
Comment (Garamotte) [inline purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py]: `date_order_str = fields.Date.today()` ?
Or `date_order_str = fields.Date.context_today()` ?

File: purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+                    po_line.price_unit = new_price
+                    po_line.purchase_request_lines = [(4, line.id)]
+            else:
+                po_line_data = self._prepare_purchase_order_line(purchase,
+                                                                 item)
+                po_line_obj.create(po_line_data)
+            res.append(purchase.id)
+
+        return {
+            'domain': "[('id','in', ["+','.join(map(str, res))+"])]",
```
Comment (Garamotte) [inline purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py:235]: `'domain': str([('id', 'in', res)]),`

File: purchase_request_to_rfq/models/purchase_order.py (hunk 1 above)
Comment (adrienpeiffer) [reply purchase_request_to_rfq/models/purchase_order.py:96]: Do you mean `self.mapped('purchase_request_lines').ids` ?

Comment (Garamotte) [reply purchase_request_to_rfq/models/purchase_order.py:96]: Oops, yes, sorry :)

**Comment by adrienpeiffer** [comment]
I added some fixes/improvements in my last commits

**Comment by JordiBForgeFlow** [comment]
@adrienpeiffer This PR from 9.0 should be moved to 10.0 https://github.com/OCA/purchase-workflow/pull/291 once it's merged.

**Comment by adrienpeiffer** [comment]
@jbeficent this is strange ... I have never seen this problem

**Comment by Garamotte** [comment]
@adrienpeiffer The `update` method allows to change the value of multiple fields at a time, but only allows a single record, so using it after `filtered` is not safe.

**Comment by JordiBForgeFlow** [comment]
@adrienpeiffer You can reproduce in runbot 9.0

**Comment by JordiBForgeFlow** [comment]
Could you cherry-pick the commits from merged https://github.com/OCA/purchase-workflow/pull/291 to this PR?

**Comment by adrienpeiffer** [comment]
@jbeficent Done !

**Comment by JordiBForgeFlow** [comment]
@adrienpeiffer can you attend to the few travis errors? Also, curious why runbot is still complaining that the purchase_request module is not existing. It has already been merged. Perhaps the PR needs rebasing?

**Comment by adrienpeiffer** [comment]
Rebased

File: purchase_request_to_rfq/models/purchase_request.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        'purchase.order.line', 'purchase_request_purchase_order_line_rel',
+        'purchase_request_line_id',
+        'purchase_order_line_id', 'Purchase Order Lines',
+        readonly=True, copy=False)
+    purchase_state = fields.Selection(
+        compute="_compute_purchase_state",
+        string="Purchase Status",
+        selection=lambda self:
+        self.env['purchase.order']._fields['state'].selection,
+        stored=True,
```
Comment (Garamotte) [inline purchase_request_to_rfq/models/purchase_request.py]: `store=True`

File: purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+                              required=True,
+                              readonly=True)
+    request_id = fields.Many2one('purchase.request',
+                                 related='line_id.request_id',
+                                 string='Purchase Request',
+                                 readonly=True)
+    product_id = fields.Many2one('product.product', string='Product')
+    name = fields.Char(string='Description', required=True)
+    product_qty = fields.Float(string='Quantity to purchase',
+                               digits_compute=dp.get_precision('Product UoS'))
```
Comment (Garamotte) [inline purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py]: `digits=dp.get_precision('Product UoS'))`

File: purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py [hunk 8]
```diff
@@ -0,0 +1,279 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016 Eficent Business and IT Consulting Services S.L.
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl-3.0).
+
+import odoo.addons.decimal_precision as dp
+from odoo import _, api, exceptions, fields, models
+from odoo.tools import DEFAULT_SERVER_DATE_FORMAT, \
+    DEFAULT_SERVER_DATETIME_FORMAT
```
Comment (Garamotte) [inline purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py]: Unused, please remove the import.

File: purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+from datetime import datetime
```
Comment (Garamotte) [inline purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py]: Unused, please remove the import.

File: purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+            if any(f not in vals for f in changed_fields):
+                obj = cls.new(vals)
+                getattr(obj, onchange_method)()
+                for field in changed_fields:
+                    vals[field] = obj._fields[field].convert_to_write(
+                        obj[field], obj)
+
+    @api.model
+    def _prepare_purchase_order_line(self, po, item):
+        po_line_obj = self.env['purchase.order.line']
```
Comment (Garamotte) [inline purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py]: Unused, please remove.

**Review by Garamotte — CHANGES_REQUESTED** [review]
Some comments about renamed field attributes and unused variables/imports.

The infinite loop that breaks the build on runbot is caused by the `digits_compute` attribute (the one in this PR, and another in the `purchse_request` module).

**Comment by Garamotte** [comment]
I opened a new PR to fix the field attribute error in `purchase_request` : https://github.com/OCA/purchase-workflow/pull/295

**Comment by sbidoul** [comment]
@adrienpeiffer could you fix the small lint errors?

```
purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py:7:1: F401 'odoo.tools.DEFAULT_SERVER_DATE_FORMAT' imported but unused
purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py:7:1: F401 'odoo.tools.DEFAULT_SERVER_DATETIME_FORMAT' imported but unused
purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py:9:1: F401 'datetime.datetime' imported but unused
purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py:153:9: F841 local variable 'po_line_obj' is assigned to but never used
```

File: purchase_request_to_rfq/views/purchase_request_view.xml [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+
+        <record id="purchase_request_line_search" model="ir.ui.view">
+            <field name="name">purchase.request.line.search</field>
+            <field name="model">purchase.request.line</field>
+            <field name="inherit_id" ref="purchase_request.purchase_request_line_search"/>
+            <field name="arch" type="xml">
+                    <field name="date_required" position="after">
+                        <filter name="purchase_state_none"
+                                string="Purchasing not started"
+                                domain="[('purchase_state','=','none')]"
```
Comment (sergio-teruel) [inline purchase_request_to_rfq/views/purchase_request_view.xml]: I think that the state never is 'none' so better do [('purchase_state', '=', False)]

**Review by sergio-teruel — CHANGES_REQUESTED** [review]

**Comment by adrienpeiffer** [comment]
@sbidoul it's done.
@sergio-teruel I just forward your commit on 9.0 to this branch

**Comment by sergio-teruel** [comment]
Oks. Thanks..

**Comment by AaronHForgeFlow** [comment]
It seems to me that you have to rebase your branch to make runbot work. The digits_compute error  was corrected in the product_qty field of the purchase request line.

File: purchase_request_to_rfq/models/purchase_request.py [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+
+class PurchaseRequestLine(models.Model):
+
+    _inherit = "purchase.request.line"
+
+    @api.multi
+    @api.depends('purchase_lines')
+    def _compute_is_editable(self):
+        super(PurchaseRequestLine, self)._compute_is_editable()
+        for rec in self.filtered(lambda p: p in self and p.purchase_lines):
```
Comment (Garamotte) [inline purchase_request_to_rfq/models/purchase_request.py:21]: Is it really required when filtering on `self` to check that the record is in `self` ?

**Comment by sbidoul** [comment]
rebased. travis and runbot are green

File: purchase_request_to_rfq/models/purchase_request.py (hunk 12 above)
Comment (JordiBForgeFlow) [reply purchase_request_to_rfq/models/purchase_request.py:21]: Hi @sylvain-garancher Not sure. What would you suggest instead?

File: purchase_request_to_rfq/models/stock.py (hunk 3 above)
Comment (JordiBForgeFlow) [reply purchase_request_to_rfq/models/stock.py:43]: @sylvain-garancher Why? purchase_line_id it is used within.

File: purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+            available_po_lines = po_line_obj.search(domain)
+            if available_po_lines:
+                po_line = available_po_lines[0]
+                new_qty = pr_line_obj._calc_new_qty(line, po_line=po_line)
+                if new_qty > po_line.product_qty:
+                    po_line.product_qty = new_qty
+                    # Leave the purchase order order to calculate its prices
+                    # itself
+                    po_line._onchange_quantity()
+                    po_line.purchase_request_lines = [(4, line.id)]
```
Comment (JordiBForgeFlow) [inline purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py:227]: This line should be executed regardless if new_qty > po_line.product_qty. \

When you create two PR lines for 1 unit for imac, which has a minimum qty of 5, now the PO line is only linked to PR 1, not to PR2. That is incorrect.

This issue also occurs in 9.0 and 8.0

File: purchase_request_to_rfq/models/purchase_request.py (hunk 12 above)
Comment (Garamotte) [reply purchase_request_to_rfq/models/purchase_request.py:21]: I suggest to replace `self.filtered(lambda p: p in self and p.purchase_lines)` by `self.filtered(lambda p: p.purchase_lines)`.

The `filtered` method loops over `self`, so `p` should always be part of `self`, isn't it ?

File: purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py (hunk 13 above)
Comment (JordiBForgeFlow) [reply purchase_request_to_rfq/wizard/purchase_request_line_make_purchase_order.py:227]: Created a separate issue #333

**Comment by LoisRForgeFlow** [comment]
This looks good! I agree with the changes proposed by @jbeficent and @sylvain-garancher though and also I propose to add this commit https://github.com/OCA/purchase-workflow/pull/346/commits/93d284bad472623ffdd3aa34bb0482ee8711e7bb to be able to add some description in the wizard and keep it in the created PO.

**Comment by LoisRForgeFlow** [comment]
Hello @adrienpeiffer , can you also add the fix made here: https://github.com/OCA/purchase-workflow/pull/371 for v9?

**Comment by aduijs** [comment]
Whats the status on this migration?

**Comment by JordiBForgeFlow** [comment]
We will merge this now. Review comments correspond to requests for new features, and I propose to crate separate PR's.

**Comment by JordiBForgeFlow** [comment]
Thanks @adrienpeiffer! This is now merged.

**Pull request merged**